Sources/Workspace/Workspace.swift

```diff
@@ -836,7 +844,10 @@ extension Workspace {
         diagnostics: DiagnosticsEngine
     ) -> [Manifest] {
         let rootManifests = packages.compactMap({ package -> Manifest? in
-            loadManifest(packagePath: package, url: package.pathString, packageKind: .root, diagnostics: diagnostics)
+            delegate?.willLoadPackage(url: package.pathString, kind: .root)
+            let manifest = loadManifest(packagePath: package, url: package.pathString, packageKind: .root, diagnostics: diagnostics)
```

Comment:
Maybe just put the delegate calls in the loadManifest method since it’s in the same class?

Reply:
I started with that, actually, but it turns out to get called multiple times for some packages, so these callbacks only wrap some of the calls to `loadPackage()`.  What I really want to do here is to define callbacks for the conceptual model that makes sense to the user, and then do whatever is needed in libSwiftPM to support that, rather than just wrap the internal calls.